### ISSUE TITLE
How to set up a mpv build and dev environment on Windows?

### ISSUE THREAD

**Issue by stax76** [mpv-player/mpv#9918]
I work on a mpv frontend for Windows called mpv.net and for that it would be helpful having a libmpv build environment and a mpv dev environment.

The libmpv build environment would help me to react to issues much quicker.

A mpv dev environment would help me to gain insight and possibly contribute.

Unfortunately, I have only little time for this project (health problem) and very little experience, so it’s a long term goal.

What I already know is the following:

- Basic C++ experience using Visual Studio.
- Very advanced Windows command line and PowerShell experience.
- git basics

What I don’t know:

- Linux, I have Ubuntu installed as dual boot and on WSL2, but haven’t much experience. I’ve started to learn the Linux command line, but I’m mostly in the beginning so far.
- I don’t know how to build C/C++ code unless it’s a Visual Studio Project.
- I don’t know vim and don’t intend to learn it because it has some characteristics that I don’t like, that is operating in different modes and having an own extension language and running in the command shell.
- I know powershell but no bash and no cmd .bat batch

What I have achieved so far is:

- I was able to build libmpv using mabs, it worked, the file size I got was about 140 MB, I probably will have to find a way of shrinking it.
- I was able to set up VS Code and MSYS2 and successfully build and debug a C hello world app.

How difficult is it and which topics to study would help me with this goal?

**Comment by Hrxn** [comment]
Not sure.. I mean, you already have MSYS2 and all the stuff needed to build mpv natively on Windows.

You've probably already seen this.. but if not: [Native Complilation (..)](https://github.com/mpv-player/mpv/blob/master/DOCS/compile-windows.md#native-compilation-with-msys2)

I don't think there is anything you'd really need besides MSYS2 and git... I mean, you can use VS Code to edit C++ just fine.

**Comment by qyot27** [comment]
I always cross-compile, which is easier to set up under native Linux, rather than WSL1 (where I remember encountering problems actually building the cross environment, although it would accept a cross-env package that was prebuilt on Linux).  I would assume WSL2 is near identical to the experience under native Linux, so the process may work much more smoothly there.

Either way, [there's also the tedious cross-compiling guide](https://github.com/qyot27/mpv/blob/extra-new/DOCS/crosscompile-mingw-tedious.txt), which goes over how to do it manually for almost every piece from a more-or-less bare Ubuntu install.  Most of it is copy-and-paste unless there's a new build issue that's cropped up in the meanwhile (it gets updated close to the same six-month schedule Ubuntu does, barring the periodic push because I wanted to catch the repo up to upstream).  There's certain problems with some of the dependencies that have arisen (like anything related to Vulkan), but it goes fairly deep down the dependency chain.  It's nearly 100% static as well, with exceptions noted where relevant.

**Comment by avih** [comment]
> it would be helpful having a libmpv build environment and a mpv dev environment.

On its own it shouldn't be too hard, but it can get quite hard depending on your needs.

> The libmpv build environment would help me to react to issues much quicker.

Much quicker than what? there are 3rd part builds, mainly shinchiro's about every week or two (source https://github.com/shinchiro/mpv-winbuild-cmake/ and binaries https://sourceforge.net/projects/mpv-player-windows/files/ ), and more recently this which provides daily automated releases based on shinchiro's setup https://github.com/zhongfly/mpv-winbuild .

> A mpv dev environment would help me to gain insight and possibly contribute.

These two needs suggest that you're not going to distribute it, i.e. it's mainly for yourself - e.g. maybe checking some recent mpv commit in practice, or trying to fiddle with [lib]mpv.

In contrast, if you intend to distribute it then it will be harder, because for it to be reasonably useful it will probably need:
- To be linked statically so that you don't end up with dozens of dependency dlls.
- Many dependencies (and dependencies of dependencies) which maybe you don't care about personally, but your users might, like vulkan, libplacebo, nvidia things, av1 in ffmpeg, encoding via libx264/5, and many more.
- If you link statically then you need all the dependency libs also ready for static linking. Most/all mingw setups don't have that by default, which means that you'll need to build all the dependencies yourself. You can look here how many dependency packages (libs) are built for every shinchiro build, each with its own recipe and caveats - https://github.com/shinchiro/mpv-winbuild-cmake/tree/master/packages .

mpv supports only mingw setups. If you must use MSVC then you're on your own.

For mingw, these are the main options:
- Use some environment which provides mingw development packages. MSYS2 is one such env. Another could be some linux distro which provides mingw packages (debian, fedora, and many more). While technically mingw in MSYS2 is cross compiling, it hides this fact so that it's almost native. But on linux it's really cross compiling, which could be easier or harder depending on $things.
- Build the mingw environment yourself (compile the mingw compiler, and everything else from scratch). This is a much more involving process but you gain control over it from A to Z.

MXE is a system to build mingw from scratch (typically on linux).

Shinchiro's setup also builds mingw from scratch (also typically on linux).

So if you want to build [lib]mpv for yourself, then you most likely want to get as many things as possible (mainly deps libraries) working out of the box and built by someone else, so that once the deps are installed then building mpv itself is a relatively quick process (2-5 minutes).

If you go this route, and you prefer to not cross compile on linux, then you should use MSYS2, install some dependencies which mpv needs (at least libass, ffmpeg, lua 5.1/5.2/jit, and their dependencies will be instaled automatically), and python, and then clone the mpv repository and run `./bootstrap` (once), and then whenever you want to rebuild mpv (e.g. after updating or editing the source files), you'll need `./waf configure && ./waf build`, and hopefully you'll get `mpv.exe` and `mpv.com` at the `build/` dir.

If you also need libmpv then you should instead use `./waf configure --enable-libmpv-shared && ./waf build`

Because all the MSYS2 mingw libs are dynamic, it means that the resulting binary (mpv or libmpv) will depend on many dlls. After you build mpv, you can list the dependency dlls with `ldd ./build/mpv.exe`.

There's likely to be more to it, but you'll need to handle it as you go, and ask questions if you get stuck.

> I’ve started to learn the Linux command line, but I’m mostly in the beginning so far.

This. You really need this. The MSYS2 shell (bash by default) is quite similar to linux.

The bottom line is that there's no simple recipe to build mpv for windows, and it really depends on your needs. It requires setup and effort and handling issues along the way, even if you use the simplest approach which is probably MSYS2.

**Comment by hooke007** [comment]
> I don't know what "mabs" is

https://github.com/m-ab-s/media-autobuild_suite
I tried the m-ab-s's full build previously, which is ultra large. It could be the main reason why his mpv is huge too.

**Comment by garoto** [comment]
I use a VirtualBox VM which I installed from a `debian-testing-amd64-netinst.iso` image (only ~350MB back in 2020), which if you don't check any options during setup, by default will give you a pretty barebones and super lean CLI-only Linux environment. Cloned and followed the instructions on shinchiro's repo README.md and had a working cross-compiling environment setup in about ~3hrs with my shitty rig.

Setup the VM with a fixed LAN IP address and access it via a win32 ssh binary via cmd.exe. Mount the "remote" box network via either SAMBA or via SSHFS, so to browse "confortably" via windows facilities like explorer.exe.

Many possibilities.

The whole VM setup at the time was occupying no more than ~1.2GB of disk space (probably reasonably less).

I was scared as well my first time, and postponed to try and replicate shinchiros/mia's cmake setup locally for the longest time, but it all worked at the end of the day. Also, the internet in 2022 have all of the answers to anything related to git, Bash, Linux, gcc, cmake, etc if you ask the right questions.

**Comment by stax76** [comment]
Having daily builds available from zhongfly should be sufficient for me, otherwise I can try building it myself.

I was now able to set up vs code correctly for mpv development and debugging.

A remaining issue I have is it takes 50 seconds to compile every time I start it, is this normal or might there be a way to speed it up?

Here is the config:

build-vs-code.sh

```
CFLAGS="-g2 -O0"
/usr/bin/python3 waf configure --disable-optimize CC=gcc.exe --check-c-compiler=gcc --prefix=$MSYSTEM_PREFIX
/usr/bin/python3 waf install
```

.vscode\launch.json

```JSON
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "debug mpv",
            "type": "cppdbg",
            "request": "launch",
            "program": "D:\\Software\\Development\\MSYS2\\mingw64\\bin\\mpv.exe",
            "args": [ "D:\\Samples\\test.vob" ],
            "stopAtEntry": false,
            "cwd": "${fileDirname}",
            "environment": [],
            "externalConsole": false,
            "MIMode": "gdb",
            "miDebuggerPath": "D:\\Software\\Development\\MSYS2\\mingw64\\bin\\gdb.exe",
            "setupCommands": [
                {
                    "description": "Enable pretty-printing for gdb",
                    "text": "-enable-pretty-printing",
                    "ignoreFailures": true
                }
            ],
            "preLaunchTask": "build mpv"
        }
    ]
}
```

.vscode\tasks.json

```JSON
{
	"version": "2.0.0",
	"tasks": [
        {
            "type": "shell",
            "label": "build mpv",
            "command": "./build-vs-code.sh",
            "args": [],
            "options": {
                "shell": {
                    "executable": "D:\\Software\\Development\\MSYS2\\msys2_shell.cmd",
                    "args": ["-defterm", "-mingw64", "-no-start", "-here", "-shell bash -c"]
                },
            },
            "group": {
                "kind": "build",
                "isDefault": true
            }
        }
    ]
}
```

**Comment by avih** [comment]
> A remaining issue I have is it takes 50 seconds to compile every time I start it, is this normal or might there be a way to speed it up?
> 
> Here is the config:
> 
> build-vs-code.sh
> 
> ```
> CFLAGS="-g2 -O0"
> /usr/bin/python3 waf configure --disable-optimize CC=gcc.exe --check-c-compiler=gcc --prefix=$MSYSTEM_PREFIX
> /usr/bin/python3 waf install
> ```

configure is a long-ish step. If it's already configured/built, and you only change some source files then you can run `waf build`, and use the same configuration (which options are enabled, which libs were detected, etc).

Also, "install" does `waf build && waf install`. It may or may not be needed, but I don't know how long it takes - I never install mpv myself.

Typically `build/mpv.exe` at the source dir is enough to run it (after `waf build`), but if you run it outside of MSYS2 (e.g. from vscode), then maybe it won't find the dlls without installing it into the mingw bin dir.

**Comment by stax76** [comment]
With `waf build` it's 11 seconds, which is fine. It's working, and I've learned something along the way. Again thanks everybody.

**Comment by kDurian** [comment]
I have been studying your mpv.net project recently and I want to understand some of the specific implementations in mpv. However, I haven't figured out how to debug mpv using VS on a Windows system yet. Could you please tell me how you achieved this?

**Comment by stax76** [comment]
I use msys2, vscode, gcc and gdb.

With Visual Studio, it might be possible with Clang, I don't know if somebody has tried it.

**Comment by miaomiao620** [comment]
> 我使用 msys2、vscode、gcc 和 gdb。
> 
> 使用Visual Studio，可能可以使用Clang，我不知道是否有人尝试过。

Could you write an article on the steps of using VS to debug mpv source code on the Windows system? Thank you very much.
